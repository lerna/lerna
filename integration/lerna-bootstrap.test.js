"use strict";

const fs = require("fs-extra");
const globby = require("globby");
const normalizePath = require("normalize-path");
const path = require("path");

const cloneFixture = require("@lerna-test/clone-fixture")(
  path.resolve(__dirname, "../commands/bootstrap/__tests__")
);
const cliRunner = require("@lerna-test/cli-runner");
const initFixture = require("@lerna-test/init-fixture")(__dirname);

describe("lerna bootstrap", () => {
  test("bootstraps all packages", async () => {
    const cwd = await initFixture("lerna-bootstrap");
    const lerna = cliRunner(cwd);

    const { stderr } = await lerna("bootstrap", "--no-ci");
    expect(stderr).toMatchSnapshot("stderr");

    // the "--silent" flag is passed to `npm run`
    const { stdout } = await lerna("run", "test", "--", "--silent");
    expect(stdout).toMatchSnapshot("stdout");

    const lockfiles = await globby(["package-*/package-lock.json"], { cwd, absolute: true });
    const [lock1, lock2, lock3] = await Promise.all(lockfiles.sort().map(fp => fs.readJson(fp)));

    expect(lock1).toMatchObject({
      name: "@integration/package-1",
      version: "1.0.0",
      dependencies: {
        pify: expect.any(Object),
        "tiny-tarball": {
          version: "1.0.0",
          optional: true,
        },
      },
    });
    expect(lock2).toMatchObject({
      name: "@integration/package-2",
      version: "1.0.0",
      dependencies: { pify: expect.any(Object) },
    });
    expect(lock3).toMatchObject({
      name: "@integration/package-3",
      version: "1.0.0",
      dependencies: { pify: expect.any(Object) },
    });
  });

  test("hoists correctly", async () => {
    const cwd = await initFixture("lerna-bootstrap");
    const lerna = cliRunner(cwd);

    const { stderr } = await lerna("bootstrap", "--no-ci", "--hoist");
    expect(stderr).toMatchSnapshot("stderr");

    // the "--silent" flag is passed to `npm run`
    const { stdout } = await lerna("run", "test", "--", "--silent");
    expect(stdout).toMatchSnapshot("stdout");

    const lockfiles = await globby(["**/package-lock.json"], {
      cwd,
      absolute: true,
      followSymlinkedDirectories: false,
    });
    const [lock3, rootLock] = await Promise.all(lockfiles.sort().map(fp => fs.readJson(fp)));

    expect(lock3).toMatchObject({
      name: "@integration/package-3",
      version: "1.0.0",
      dependencies: { pify: expect.any(Object) },
    });
    expect(rootLock).toMatchObject({
      name: "integration",
      version: "0.0.0",
      dependencies: {
        pify: expect.any(Object),
        "tiny-tarball": {
          version: "1.0.0",
          // root hoist does not preserve optional
        },
      },
    });
    expect(rootLock).not.toHaveProperty("dependencies.tiny-tarball.optional");
  });

  test("--npm-client yarn", async () => {
    const cwd = await initFixture("lerna-bootstrap");
    const lerna = cliRunner(cwd);

    const { stderr } = await lerna("bootstrap", "--npm-client", "yarn");
    expect(stderr).toMatchSnapshot("stderr");

    const lockfiles = await globby(["package-*/yarn.lock"], { cwd });
    expect(lockfiles.sort().map(fp => normalizePath(fp))).toEqual([
      "package-1/yarn.lock",
      "package-2/yarn.lock",
      "package-3/yarn.lock",
    ]);

    // the "--silent" flag is passed to `npm run`
    const { stdout } = await lerna("run", "test", "--", "--silent");
    expect(stdout).toMatchSnapshot("stdout");
  });

  test("--ci", async () => {
    const { cwd } = await cloneFixture("ci");
    const lerna = cliRunner(cwd);

    const { stderr } = await lerna("bootstrap", "--ci");
    expect(stderr).toMatchSnapshot("stderr");

    // the "--silent" flag is passed to `npm run`
    const { stdout } = await lerna("run", "test", "--", "--silent");
    expect(stdout).toMatchSnapshot("stdout");
  });
});
