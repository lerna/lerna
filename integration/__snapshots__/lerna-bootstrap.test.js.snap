// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lerna bootstrap --ci: stderr 1`] = `
lerna info version __TEST_VERSION__
lerna info Bootstrapping 1 packages
lerna info Installing external dependencies
lerna info Symlinking packages and binaries
lerna success Bootstrapped 1 packages
`;

exports[`lerna bootstrap --ci: stdout 1`] = `package-1`;

exports[`lerna bootstrap --npm-client yarn: stderr 1`] = `
lerna info version __TEST_VERSION__
lerna info Bootstrapping 4 packages
lerna info lifecycle package-4@1.0.0~preinstall: package-4@1.0.0
lerna info Installing external dependencies
lerna info Symlinking packages and binaries
lerna info lifecycle @integration/package-3@1.0.0~postinstall: @integration/package-3@1.0.0
lerna info lifecycle @integration/package-2@1.0.0~prepublish: @integration/package-2@1.0.0
lerna info lifecycle @integration/package-1@1.0.0~prepare: @integration/package-1@1.0.0
lerna success Bootstrapped 4 packages
`;

exports[`lerna bootstrap --npm-client yarn: stdout 1`] = `
package-1
package-2
cli package-2 OK
package-3 cli1 OK
package-3 cli2 package-2 OK
`;

exports[`lerna bootstrap bootstraps all packages: stderr 1`] = `
lerna info version __TEST_VERSION__
lerna info Bootstrapping 4 packages
lerna info lifecycle package-4@1.0.0~preinstall: package-4@1.0.0
lerna info Installing external dependencies
lerna info Symlinking packages and binaries
lerna info lifecycle @integration/package-3@1.0.0~postinstall: @integration/package-3@1.0.0
lerna info lifecycle @integration/package-2@1.0.0~prepublish: @integration/package-2@1.0.0
lerna info lifecycle @integration/package-1@1.0.0~prepare: @integration/package-1@1.0.0
lerna success Bootstrapped 4 packages
`;

exports[`lerna bootstrap bootstraps all packages: stdout 1`] = `
package-1
package-2
cli package-2 OK
package-3 cli1 OK
package-3 cli2 package-2 OK
`;

exports[`lerna bootstrap hoists correctly: stderr 1`] = `
lerna info version __TEST_VERSION__
lerna info Bootstrapping 4 packages
lerna info lifecycle package-4@1.0.0~preinstall: package-4@1.0.0
lerna WARN EHOIST_PKG_VERSION "@integration/package-3" package depends on pify@^1.0.0, which differs from the hoisted pify@^2.0.0.
lerna info Installing external dependencies
lerna info hoist Installing hoisted dependencies into root
lerna info hoist Pruning hoisted dependencies
lerna info hoist Finished pruning hoisted dependencies
lerna info hoist Finished bootstrapping root
lerna info Symlinking packages and binaries
lerna info lifecycle @integration/package-3@1.0.0~postinstall: @integration/package-3@1.0.0
lerna info lifecycle @integration/package-2@1.0.0~prepublish: @integration/package-2@1.0.0
lerna info lifecycle @integration/package-1@1.0.0~prepare: @integration/package-1@1.0.0
lerna success Bootstrapped 4 packages
`;

exports[`lerna bootstrap hoists correctly: stdout 1`] = `
package-1
package-2
cli package-2 OK
package-3 cli1 OK
package-3 cli2 package-2 OK
`;
